### PR TITLE
Separate global optimization constraints for local SLAM and odometry

### PR DESCRIPTION
**Pull request by MichaelGrupp**
Implements [RFC-0018](https://github.com/googlecartographer/rfcs/blob/master/text/0018-odometry-use-in-optimization-problem.md)

### CODE REVIEW

**Comment by MichaelGrupp** [comment]
A note on follow-ups to introduce the new parameter names instead of `consecutive_node_{rotation, translation}_weight` in othe other repositories:

* `cartographer_ros`: not required, the backpack configs don't override the default parameters
* `cartographer_{fetch, toyota_hsr, turtlebot}`: `cartographer_fetch` overrides the default parameters, but I'd prefer if someone from your team would change the configuration since I didn't do the initial tuning.
  ```bash
  $ ag consecutive_node_

  cartographer_fetch/cartographer_fetch/configuration_files/grpc_freight_server.lua
  23:POSE_GRAPH.optimization_problem.consecutive_node_translation_weight = 1e3
  24:POSE_GRAPH.optimization_problem.consecutive_node_rotation_weight = 1e2

  cartographer_fetch/cartographer_fetch/configuration_files/freight.lua
  56:POSE_GRAPH.optimization_problem.consecutive_node_translation_weight = 1e3
  57:POSE_GRAPH.optimization_problem.consecutive_node_rotation_weight = 1e2
  ```

File: cartographer/mapping/internal/2d/pose_graph/optimization_problem_2d.cc [hunk 1]
```diff
@@ -273,12 +273,27 @@ void OptimizationProblem2D::Solve(
         continue;
       }
 
-      const transform::Rigid3d relative_pose =
-          ComputeRelativePose(trajectory_id, first_node_data, second_node_data);
+      // Add a relative pose constraint based on the odometry (if available).
+      std::unique_ptr<transform::Rigid3d> relative_odometry =
+          OdometryBetween(trajectory_id, first_node_data, second_node_data);
```
Comment (cschuet) [inline cartographer/mapping/internal/2d/pose_graph/optimization_problem_2d.cc]: usually we try to start function names with a verb unless they are simple accessors. How about `CalculateOdometryBetweenNodes()`?

File: cartographer/mapping/internal/2d/pose_graph/optimization_problem_2d.cc [hunk 2]
```diff
@@ -273,12 +273,27 @@ void OptimizationProblem2D::Solve(
...
+          transform::Embed3D(first_node_data.initial_pose.inverse() *
+                             second_node_data.initial_pose);
       problem.AddResidualBlock(
-          SpaCostFunction2D::CreateAutoDiffCostFunction(Constraint::Pose{
-              relative_pose, options_.consecutive_node_translation_weight(),
-              options_.consecutive_node_rotation_weight()}),
+          SpaCostFunction2D::CreateAutoDiffCostFunction(
+              Constraint::Pose{relative_local_slam_pose,
+                               options_.local_slam_pose_translation_weight(),
+                               options_.local_slam_pose_rotation_weight()}),
```
Comment (cschuet) [inline cartographer/mapping/internal/2d/pose_graph/optimization_problem_2d.cc]: Is Ceres smart enough not to include constraints in the optimization problem that have a weight of 0? So I assume we don't need to special case this.

Comment (MichaelGrupp) [reply cartographer/mapping/internal/2d/pose_graph/optimization_problem_2d.cc]: Hmm... you mean if it still works if someone sets the weights to zero? Not sure.

File: cartographer/mapping/internal/2d/pose_graph/optimization_problem_2d.cc (hunk 1 above)
Comment (MichaelGrupp) [reply cartographer/mapping/internal/2d/pose_graph/optimization_problem_2d.cc]: > CalculateOdometryBetweenNodes()

That's what we do, sounds good.

File: cartographer/mapping/internal/2d/pose_graph/optimization_problem_2d.cc (hunk 2 above)
Comment (MichaelGrupp) [reply cartographer/mapping/internal/2d/pose_graph/optimization_problem_2d.cc]: As far as I can tell, the weights are not known to Ceres, they are applied in our custom cost functor [here](https://github.com/googlecartographer/cartographer/blob/master/cartographer/mapping/internal/2d/pose_graph/spa_cost_function_2d.h#L51), specifically [here](https://github.com/googlecartographer/cartographer/blob/master/cartographer/mapping/internal/pose_graph/cost_helpers_impl.h#L42).

File: cartographer/mapping/internal/2d/pose_graph/optimization_problem_2d.cc [hunk 3]
```diff
@@ -273,12 +273,28 @@ void OptimizationProblem2D::Solve(
...
-      const transform::Rigid3d relative_pose =
-          ComputeRelativePose(trajectory_id, first_node_data, second_node_data);
+      // Add a relative pose constraint based on the odometry (if available).
+      std::unique_ptr<transform::Rigid3d> relative_odometry =
+          CalculateOdometryBetweenNodes(trajectory_id, first_node_data,
+                                        second_node_data);
+      if (relative_odometry != nullptr) {
+        problem.AddResidualBlock(
+            SpaCostFunction2D::CreateAutoDiffCostFunction(Constraint::Pose{
+                (*relative_odometry), options_.odometry_translation_weight(),
```
Comment (gaschler) [inline cartographer/mapping/internal/2d/pose_graph/optimization_problem_2d.cc]: are these brackets necessary?

File: cartographer/mapping/internal/2d/pose_graph/optimization_problem_2d.cc [hunk 4]
```diff
@@ -325,7 +341,8 @@ std::unique_ptr<transform::Rigid3d> OptimizationProblem2D::InterpolateOdometry(
           .transform);
 }
 
-transform::Rigid3d OptimizationProblem2D::ComputeRelativePose(
+std::unique_ptr<transform::Rigid3d>
```
Comment (gaschler) [inline cartographer/mapping/internal/2d/pose_graph/optimization_problem_2d.cc]: The return type could be common::optional but I think consistency with the other  function has priority here.

Comment (MichaelGrupp) [reply cartographer/mapping/internal/2d/pose_graph/optimization_problem_2d.cc]: How you like.

File: cartographer/mapping/internal/2d/pose_graph/optimization_problem_2d.cc (hunk 3 above)
Comment (MichaelGrupp) [reply cartographer/mapping/internal/2d/pose_graph/optimization_problem_2d.cc]: Removing. I keep the ones in the multiplications though, they add better readability.

File: cartographer/mapping/internal/2d/pose_graph/optimization_problem_2d.cc (hunk 2 above)
Comment (cschuet) [reply cartographer/mapping/internal/2d/pose_graph/optimization_problem_2d.cc]: Yes, that is what I was wondering about.

**Comment by cschuet** [comment]
@wally-the-cartographer merge

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user cschuet. Merge queue now has a length of 1.

**Pull request merged**